File: structural/bridge/typescript/refined-abstraction.ts
import { RemoteControl } from "./abstraction";

export class BasicRemote extends RemoteControl {
	togglePower(): void {
		if (this.device.isOn()) {
			this.device.off();
		} else {
			this.device.on();
		}
	}
}

export class AdvancedRemote extends RemoteControl {
	togglePower(): void {
		if (this.device.isOn()) {
			this.device.off();
		} else {
			this.device.on();
		}
	}

	mute(): void {
		console.log('Device is muted');
	}
}

File: structural/bridge/python/refined_abstraction.py
from abstraction import RemoteControl


class BasicRemote(RemoteControl):
    def toggle_power(self):
        if self._device.is_on():
            self._device.off()
        else:
            self._device.on()


class AdvancedRemote(RemoteControl):
    def toggle_power(self):
        if self._device.is_on():
            self._device.off()
        else:
            self._device.on()

    def mute(self):
        print("Device is muted")
